### PR TITLE
Introduce Read-Only Scoped Transactions

### PR DESCRIPTION
**Pull request by vjeko**
Introduce a new transaction type (called ScopedTxn) whose lifespan is defined by the lifespan of the ScopenTxn instance, and is not tied the thread local storage.

When defining a scoped transaction, the client has to specify a set of tables he or she is interested in during the initialization, which will effectively create a snapshot transaction. The underlying tables are not tied to MVO cache and are not eligible for eviction.

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java [hunk 1]
```diff
@@ -239,41 +265,11 @@ public TxnContext txn(@Nonnull final String namespace, IsolationLevel isolationL
     public ScopedTransaction scopedTxn(
             @Nonnull final String namespace, IsolationLevel isolationLevel,
             Table<?, ?, ?>... tables) {
-        try {
-            return IRetry.build(IntervalRetry.class, () -> {
```
Comment (hisundar) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java:243]: Hi @vjeko - apologies if I missed this, but where is this retry moved to? I don't see it in the cnostructor around txBeginInternal right? The MVCC layer can throw a `TrimmedException` which we need to catch and retry when getting a scoped transaction snapshot right?
This was actually an issue that we found during our scale demo testing

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedCorfuTable.java [hunk 2]
```diff
@@ -5,14 +5,16 @@
...
 import org.corfudb.runtime.object.ICorfuSMRUpcallTarget;
+import org.corfudb.runtime.object.transactions.Transaction;
 
 import javax.annotation.Nonnull;
 import java.util.Map;
-import org.corfudb.runtime.object.ICorfuSMRProxyMetadata;
-
+import java.util.Optional;
 import java.util.Set;
+import java.util.function.Consumer;
```
Comment (hisundar) [inline runtime/src/main/java/org/corfudb/runtime/collections/PersistedCorfuTable.java:17]: Can we remove the unused imports please? Best is if you can do that cool trick where you optimize all the imports :) thanks

**Review by hisundar — COMMENTED** [review]
Thanks so much @vjeko - this change makes it easy to consume since there are fewer interface changes, just a few comments from me.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java [hunk 3]
```diff
@@ -93,23 +93,53 @@ private Timestamp getTimestamp() {
...
+        ArrayList<Table> federatedTables = new ArrayList<>();
+        runtime.getTableRegistry().getAllOpenTables().forEach(t -> {
+            String tableNameWithoutNs = StringUtils.substringAfter(t.getFullyQualifiedTableName(),
+                    t.getNamespace() + "$");
+            CorfuStoreMetadata.TableName tableName = CorfuStoreMetadata.TableName.newBuilder()
+                    .setNamespace(t.getNamespace())
+                    .setTableName(tableNameWithoutNs).build();
+            CorfuRecord<CorfuStoreMetadata.TableDescriptors, CorfuStoreMetadata.TableMetadata> tableRecord =
+                    runtime.getTableRegistry().getRegistryTable().get(tableName);
+            if (tableRecord.getMetadata().getTableOptions().getIsFederated()) {
```
[the remark this answers is not in the document]
Comment (shama358) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java:116]: Hey Vjeko, incase you have to push a new commit to this PR, can you please include this line above L116?
"TODO V2: change the condition appropriately for routing Queue model"

If not, no worries, I'll add it after we merge this.

**Review by shama358 — APPROVED** [review]
Mostly looks good. 
Okay to merge after addressing some of the codacy errors.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java:116]: Actually the plan is to get a smaller subset of tables that will be part of the data read for the full snapshot sync and not have this method at all.

**Review by hisundar — APPROVED** [review]
Change looks good to me and we can merge it once most, if not all the codacy error can be addressed. thanks

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java [hunk 4]
```diff
@@ -93,23 +93,53 @@ private Timestamp getTimestamp() {
...
+                federatedTables.add(t);
+            }
+        });
+        return new ScopedTransaction(runtime, namespace,
```
Comment (hisundar) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java:120]: Hi @vjeko - sorry if I missed it but I don't see the retry logic for the possible transient failure that can come from the corfu's implementation of MVCC? Can you please share where the scope transaction creation is retried?

**Review by hisundar — COMMENTED** [review]
Sorry I just noticed something while picking up this change and wanted to get clarification before this is merged?

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java (hunk 4 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: Looks like I missed that path. Fixed.

**Review by hisundar — APPROVED** [review]

**Pull request merged**